### PR TITLE
Implementing persistent collectives

### LINKED ISSUE #4876: mpi4: persistent collectives
## Standard
link: https://github.com/mpi-forum/mpi-issues/issues/25
pdf: https://github.com/mpi-forum/mpi-standard/files/2352702/mpi32-report-ticket25-barcelona-vote-sep2018.pdf

## Action Items
* 21 new APIs
* Basic implementation
  ** Re-use nonblocking collective algos w/o freeing request at completion
  ** Modification to Requests, Schedules, MPI_START[ALL], MPI_REQUEST_FREE

* Advanced implementation
  ** Utilize the init phase to optimize some collective operations

## Pull requests
PR #4859 - coll: persistent bcast implementation

### PR DESCRIPTION
**Pull request by yfguo**
## Pull Request Description

Some changes are cherry-picked from Taru's branch updated for automated interface generation. All codes are refactor given that nonblocking coil calls schedule functions directly.

Add minor fixes to the problems in Bcast_init that slipped through.

A change to the python code to handle the function name.

## Expected Impact
Fixes #4876 
## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [ ] Remove xfail from the test suite when fixing a test
* [ ] Commits are self-contained and do not do two things at once
* [ ] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [ ] Passes whitespace checkers
* [ ] Passes warning tests
* [ ] Passes all tests
* [ ] Add comments such that someone without knowledge of the code could understand
* [ ] You or your company has a signed contributor's agreement on file with Argonne
* [ ] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by yfguo** [comment]
test:mpich/ch4/ofi

**Comment by yfguo** [comment]
test:mpich/ch4/ofi

**Comment by yfguo** [comment]
test:mpich/ch4/ofi

**Comment by yfguo** [comment]
test:mpich/ch4/ofi

**Comment by yfguo** [comment]
test:mpich/ch4/ofi
test:mpich/ch3/ofi

**Comment by yfguo** [comment]
@hzhou This persistent collective PR is ready for review. Most of the persistent collective feature are here implemented using the Gentran (TSP) algorithm. There are a few of them (barrier_init, alltoallv, alltoallw, etc) that does not have a TSP algorithm, therefore are not included here. I will create a separate PR adding the sched support for non-TSP algorithms.

**Comment by yfguo** [comment]
test:mpich/ch4/ofi
test:mpich/ch3/ofi

File: src/mpi/coll/allreduce/allreduce_init.c [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+                                           count >= nranks, mpi_errno,
+                                           "Pallreduce gentran_recexch_reduce_scatter_recexch_allgatherv cannot be applied.\n");
+            /*  This algorithm will work for commutative operations and if the count is
+             * bigger than total number of ranks. If it not commutative or if the count < nranks,
+             * MPIR_Iallreduce_sched algorithm will be run */
+            mpi_errno =
+                MPIR_TSP_Iallreduce_sched_intra_recexch_reduce_scatter_recexch_allgatherv
+                (sendbuf, recvbuf, count, datatype, op, comm_ptr,
+                 MPIR_CVAR_IALLREDUCE_RECEXCH_KVAL, sched);
+            break;
```
Comment (hzhou) [inline src/mpi/coll/allreduce/allreduce_init.c:101]: It really bothers me that we are duplicating these codes. If you added the `sched` support, then I think you can refactor the code and avoid this duplication, right?

**Review by hzhou — CHANGES_REQUESTED** [review]

File: src/mpi/coll/allreduce/allreduce_init.c (hunk 1 above)
Comment (yfguo) [reply src/mpi/coll/allreduce/allreduce_init.c:101]: I agree. It is getting worse when non-gentran algorithms are added.

**Comment by yfguo** [comment]
test:mpich/ch4/ofi
test:mpich/ch3/ofi

**Comment by yfguo** [comment]
test:mpich/ch4/ofi
test:mpich/ch3/ofi

**Comment by yfguo** [comment]
test:mpich/ch4/ofi
test:mpich/ch3/ofi

**Comment by yfguo** [comment]
test:mpich/ch4/ofi
test:mpich/ch3/ofi

**Comment by hzhou** [comment]
@yfguo I just discovered that previous PR #5266 had a bug that prevents proper sched reset, resulting the skipping the sched execution beyond first round. I have fixed it in PR #5311.

If you rebase on the #5311, you will find test failures in this PR due to not handling the temporary buffers and states correctly. All buffers allocated in nonblocking collectives need be allocated using `MPIR_Sched_alloc_state` so they can remain valid between multiple `MPI_Start`.

**Comment by yfguo** [comment]
test:mpich/ch4/ofi
test:mpich/ch3/ofi

**Comment by yfguo** [comment]
@hzhou I have pushed the fix to the python script. Can you take another look? Thanks.

**Comment by hzhou** [comment]
> 
> 
> @hzhou I have pushed the fix to the python script. Can you take another look? Thanks.

Can you rebase to `main` and re-run tests?

**Comment by yfguo** [comment]
test:mpich/ch4/ofi
test:mpich/ch3/ofi

File: test/mpi/coll/testlist.in [hunk 2]
```diff
@@ -176,3 +176,22 @@ p_bcast 4 strict=false
...
+p_scatterv 4 strict=false
+p_alltoallv 10 strict=false
+p_alltoallw 10 strict=false
+p_exscan 5 strict=false
+p_barrier 2 strict=false
+p_neighb_allgather 4 strict=false
+p_neighb_allgatherv 4 strict=false
+p_neighb_alltoall 4 strict=false
+p_neighb_alltoallv 4 strict=false
+p_neighb_alltoallw 4 strict=false
```
Comment (hzhou) [inline test/mpi/coll/testlist.in:197]: Missing  `p_allgather` and `p_allgatherv`.

File: test/mpi/coll/p_allred.c [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    int one = 1, two = 2, psum = 0, sum = 0;
+
+    MTest_Init(&argc, &argv);
+    MPI_Comm_size(MPI_COMM_WORLD, &size);
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+    MPI_Info_create(&info);
+    assert(size == 2);
+    MPI_Allreduce_init(&one, &psum, 1, MPI_INT, MPI_SUM, MPI_COMM_WORLD, info, &request);
+    MPI_Start(&request);
+    MPI_Wait(&request, MPI_STATUS_IGNORE);
```
Comment (hzhou) [inline test/mpi/coll/p_allred.c:25]: Need to test multiple rounds of `Start` and `Wait`.

**Review by hzhou — CHANGES_REQUESTED** [review]

**Comment by yfguo** [comment]
test:mpich/ch4/ofi
test:mpich/ch3/ofi

Forgot to trigger the tests last time [:sigh]

**Comment by yfguo** [comment]
test:mpich/ch4/ofi
test:mpich/ch3/ofi

File: test/mpi/coll/testlist.in [hunk 4]
```diff
@@ -1,5 +1,5 @@
 allred 4 arg=-count=10
-allred 7 arg=-count=10
```
Comment (hzhou) [inline test/mpi/coll/testlist.in:2]: type

File: test/mpi/coll/testlist.in [hunk 5]
```diff
@@ -176,3 +176,24 @@ p_bcast 4 strict=false
 p_bcast2 8 strict=false
 
 coll_large 8
+p_allred 2 strict=false
+p_red 4 strict=false
+p_alltoall 8 strict=false
+p_allgatherg10 strict=false
```
Comment (hzhou) [inline test/mpi/coll/testlist.in:182]: typo (your keyboard has issues :) )

Comment (yfguo) [reply test/mpi/coll/testlist.in:182]: Yes, embarrasingly. The new keyboard firmware has some timing issues.

**Comment by yfguo** [comment]
test:mpich/ch4/ofi
test:mpich/ch3/ofi

The typos are fixed.

**Comment by yfguo** [comment]
test:mpich/ch4/ofi
test:mpich/ch3/ofi

**Comment by yfguo** [comment]
Make it draft untill fix.

**Comment by hzhou** [comment]
@yfguo For your reference -- https://github.com/pmodels/mpich/pull/5266/commits/7272bc0d5005dcfe5a7797e49f4efac669052089

**Comment by yfguo** [comment]
> @yfguo For your reference -- [7272bc0](https://github.com/pmodels/mpich/commit/7272bc0d5005dcfe5a7797e49f4efac669052089)

Thanks!

**Comment by yfguo** [comment]
test:mpich/ch4/ofi
test:mpich/ch3/ofi

**Comment by yfguo** [comment]
test:mpich/ch4/ofi
test:mpich/ch3/ofi

**Comment by yfguo** [comment]
test:mpich/ch4/ofi
test:mpich/ch3/ofi

**Comment by yfguo** [comment]
Gentran Tree algorithm for persistent allreduce seems broken. Investigating this tomorrow. Everything else is working normally.

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by yfguo** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by yfguo** [comment]
@hzhou  made some change on top of your fix. 1. making sure every persistent collective test checks the results. 2. cleanup whitespace. 3. found same problem in ireduce gentran tree algorithm and fixed (found after fix p_red).

**Comment by hzhou** [comment]
@yfguo How about squash some commits? For example, all the add test commits can be squashed together and all the "implement persistent ..." can be squashed together. Squashing the boilerplates together can help review the fixes and critical changes.

**Comment by yfguo** [comment]
> @yfguo How about squash some commits? For example, all the add test commits can be squashed together and all the "implement persistent ..." can be squashed together. Squashing the boilerplates together can help review the fixes and critical changes.

Yeah. I think we can squash by topics:
0. Python script fix for pcoll
1. Adding pcoll impl code
2. Making tmp buf persistent
3. TSP algorithm fixes
4. Adding tests
5. Etc

**Comment by yfguo** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by yfguo** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Review by hzhou — APPROVED** [review]
LGTM.

**Comment by yfguo** [comment]
Rebased and ready for merge one the tests passed

**Pull request merged**